frontend/test/lit/test_autograph.py

```diff
@@ -0,0 +1,370 @@
+# Copyright 2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Unit tests for the AutoGraph source-to-source transformation feature."""
+
+# RUN: %PYTHON %s | FileCheck %s
+
+from catalyst import qjit
+from catalyst.autograph import AutoGraphError, autograph, print_code
+
+
+# CHECK-LABEL: def if_simple
+@autograph
+def if_simple(x: float):
+    """Test a simple conditional with a single branch."""
+
+    # CHECK:   def if_body():
+    # CHECK:       pass
+    if x < 3:
+        pass
+    # CHECK:   def else_body():
+    # CHECK:       pass
+
+    # CHECK:   ag__.if_stmt(x < 3, if_body, else_body, get_state, set_state, (), 0)
+
+    return x
+
+
+print_code(if_simple)
+
+# -----
+
+
+# CHECK-LABEL: def if_else
+@autograph
+def if_else(x: float):
+    """Test a simple conditional with two branches."""
+
+    # CHECK:   def if_body():
+    # CHECK:       pass
+    if x < 3:
+        pass
+    # CHECK:   def else_body():
+    # CHECK:       pass
+    else:
+        pass
+
+    # CHECK:   ag__.if_stmt(x < 3, if_body, else_body, get_state, set_state, (), 0)
+
+    return x
+
+
+print_code(if_else)
+
+# -----
+
+
+# CHECK-LABEL: def if_assign
+@autograph
+def if_assign(x: float):
+    """Test a conditional creates a new variable."""
+
+    # CHECK:   def if_body():
+    # CHECK:       nonlocal y
+    # CHECK:       y = 4
+    if x < 3:
+        y = 4
+    # CHECK:   def else_body():
+    # CHECK:       nonlocal y
+    # CHECK:       y = 5
+    else:
+        y = 5
+
+    # CHECK:   ag__.if_stmt(x < 3, if_body, else_body, get_state, set_state, ('y',), 1)
+
+    # CHECK:   return y
+    return y
+
+
+print_code(if_assign)
+
+# -----
+
+
+try:
+
+    @qjit  # needed to trigger Catalyst type checks during tracing
+    @autograph
+    def if_assign_type_mismatch(x: float):
+        """Verify error from a conditional that doesn't produce the same type across branches."""
```

Comment:
Would it be desirable and/or a lot of work to perform any kind of automatic casting in these instances (like having `(float, int) -> float`)?

Reply:
Totally agree, I've raised this issue before as an improvement to our control flow and I believe it is on the roadmap.
But this is independent of the autograph feature. The test is really only here because I thought of this a potential scenario when I first defined the desired behaviour (see [first commit](https://github.com/PennyLaneAI/catalyst/pull/235/commits/d323848abca372ddbdb866f84a4716e8ee5fd8e7)).